File: rust/slopdesk-video/src/window_placement.rs
//! Where a remoted window goes when it is parked on the virtual display.
//!
//! The host puts the captured window on a `HiDPI` virtual display so it renders at a real 2×
//! backing.
//! macOS CROPS a window that overhangs its display, so an oversized one has to be shrunk before it
//! is moved, and after the move the achieved size has to be checked: an app that refuses or clamps
//! the resize leaves a window still overhanging, which must not be reported as a successful park —
//! the capture crop would exceed the framebuffer and the client's input mapping would desync.
//!
//! Both rules are arithmetic on six numbers, and the caller's window-system semantics stay with the
//! caller: a display's extents arrive ALREADY standardised (`CGRect.width` returns `|size|`) while
//! a window's size arrives RAW (`CGSize.width` is a stored field), so the clamp here is
//! deliberately asymmetric and this module never takes an absolute value of its own.
//!
//! The ordered comparisons are spelled as they are for one reason: a NaN. `display < window` is
//! false for a NaN operand, so the window value passes through — where a min that "ignores NaN"
//! would hand back the display's. The golden corpus pins the answers as bit patterns precisely so a
//! port cannot quietly choose the other one.

/// The half-point tolerance both rules carry.
///
/// Exactly representable, and added to values that are, so it introduces no rounding of its own. It
/// exists so floating-point equality does not order a no-op resize, or fail a window that fits.
pub const TOLERANCE: f64 = 0.5;

/// Where a window goes, at what size, and whether it has to be resized to get there.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Placement {
    /// The display origin the window is moved to, verbatim — including the negative coordinates of
    /// a display placed left of or above the main one.
    pub origin_x: f64,
    /// The vertical half of that origin.
    pub origin_y: f64,
    /// The width to resize to: the window's own, or the display's where the window is wider.
    pub width: f64,
    /// The height to resize to, by the same rule.
    pub height: f64,
    /// Whether the clamp actually shrank the window by more than the tolerance. A window that
    /// already fits is moved without a resize, so an app that dislikes being resized is not asked.
    pub needs_resize: bool,
}

/// Clamps a window to a display — DOWN only, never enlarging — and places it at the display origin.
///
/// `display_width` / `display_height` are the display's STANDARDISED extents and
/// `window_width` / `window_height` the window's RAW size, as the caller's window system defines
/// each; this function compares them as given.
#[must_use]
pub fn place(
    window_width: f64,
    window_height: f64,
    display_x: f64,
    display_y: f64,
    display_width: f64,
    display_height: f64,
) -> Placement {
    // The ordered ternary, NOT a NaN-ignoring minimum: a NaN window size stays a NaN here, and the
    // vectors pin that.
    let width = if display_width < window_width {
        display_width
    } else {
        window_width
    };
    let height = if display_height < window_height {
        display_height
    } else {
        window_height
    };
    // The tolerance is added to the CLAMPED extent and compared against the RAW window, each step
    // its own operation. There is no multiply to fuse with, but the comparison is a half-point
    // predicate the corpus pins, so it stays spelled exactly this way.
    let needs_width = width + TOLERANCE < window_width;
    let needs_height = height + TOLERANCE < window_height;
    Placement {
        origin_x: display_x,
        origin_y: display_y,
        width,
        height,
        needs_resize: needs_width || needs_height,
    }
}

/// Whether a size fits inside bounds, within the tolerance.
///
/// Read AFTER the move, against the size the window actually achieved: the plan said what to ask
/// for, and this says whether the app complied. A NaN fails, which is the answer that rolls the
/// window back rather than parking it on a size nothing can crop correctly.
#[must_use]
pub fn fits(width: f64, height: f64, bounds_width: f64, bounds_height: f64) -> bool {
    width <= bounds_width + TOLERANCE && height <= bounds_height + TOLERANCE
}

#[cfg(test)]
mod tests {
    #![expect(
        clippy::float_cmp,
        reason = "the fixtures are exact binary fractions, and passing a value through UNCHANGED is the \
                  property under test"
    )]

    use super::{fits, place};

    #[test]
    fn a_window_that_already_fits_is_moved_without_being_resized() {
        let plan = place(800.0, 600.0, 0.0, 0.0, 1920.0, 1080.0);
        assert_eq!((plan.width, plan.height), (800.0, 600.0));
        assert!(!plan.needs_resize);
    }

    #[test]
    fn an_oversized_window_is_shrunk_to_the_display_before_it_crosses() {
        let plan = place(2560.0, 1600.0, 0.0, 0.0, 1920.0, 1080.0);
        assert_eq!((plan.width, plan.height), (1920.0, 1080.0));
        assert!(plan.needs_resize, "macOS would crop it otherwise");
    }

    #[test]
    fn the_display_origin_crosses_verbatim_including_a_negative_one() {
        let plan = place(400.0, 300.0, -1920.0, -240.0, 1920.0, 1080.0);
        assert_eq!((plan.origin_x, plan.origin_y), (-1920.0, -240.0));
    }

    #[test]
    fn a_shrink_inside_the_tolerance_is_not_worth_asking_an_app_for() {
        let plan = place(1920.25, 1080.0, 0.0, 0.0, 1920.0, 1080.0);
        assert_eq!(plan.width, 1920.0, "the clamp still applies");
        assert!(!plan.needs_resize, "but a quarter point is not a resize");
        assert!(place(1921.0, 1080.0, 0.0, 0.0, 1920.0, 1080.0).needs_resize);
    }

    #[test]
    fn a_nan_window_size_passes_through_rather_than_becoming_the_display() {
        let plan = place(f64::NAN, 600.0, 0.0, 0.0, 1920.0, 1080.0);
        assert!(plan.width.is_nan(), "an ordered compare keeps the NaN");
        assert!(
            !plan.needs_resize,
            "and every comparison against it is false, so nothing is asked for"
        );
    }

    #[test]
    fn a_window_fits_when_it_overhangs_by_less_than_the_tolerance() {
        assert!(fits(1920.0, 1080.0, 1920.0, 1080.0), "exact fits");
        assert!(
            fits(1920.25, 1080.0, 1920.0, 1080.0),
            "a quarter point over still fits"
        );
        assert!(!fits(1921.0, 1080.0, 1920.0, 1080.0));
        assert!(
            !fits(f64::NAN, 1080.0, 1920.0, 1080.0),
            "an unreadable size is rolled back, not parked"
        );
    }
}

File: Sources/SlopDeskVideoHost/WindowPlacement.swift
#if os(macOS)
import CoreGraphics
import CSlopDeskFFI
import Foundation
import OSLog

/// The Swift face of `rust/slopdesk-video`'s `window_placement` (feature #1): decide where and at
/// what size to move a window fully onto a display. The AX side effects live in ``WindowPlacement``.
///
/// What stays HERE is what CoreGraphics defines, and only that:
///   * `CGRect.width`/`.height` STANDARDIZE — they return `|size|` (always ≥ 0) — so the display
///     extents are read through them and cross already standardized.
///   * `CGSize.width`/`.height` are RAW stored fields (NOT standardized), so the window operand
///     (and the `size` arg to ``fits``) crosses verbatim, never abs'd. The clamp is therefore
///     asymmetric, and the far side is told which side it is holding rather than deciding.
///   * `CGRect.origin` is the RAW stored origin (NOT standardized), so the origin crosses verbatim —
///     including negative coordinates from a display placed left of / above the main display.
///
/// What moved is the arithmetic: the ordered comparison that keeps a NaN where a NaN-ignoring
/// minimum would swallow it, and the ½-pt tolerance predicate. `golden/golden_vectors.json` pins
/// both as bit patterns, replayed by `WindowPlacementGoldenVectorTests` here and by
/// `every_pinned_placement_puts_the_window_where_swift_put_it` on the far side.
public enum WindowPlacementMath {
    /// Clamp `windowSize` to `displayBounds` (resize DOWN only if larger — never enlarge) and place
    /// at the display's top-left origin. macOS crops a window that overhangs a display, so an
    /// oversized window must be shrunk before the move.
    public static func placement(windowSize: CGSize, displayBounds: CGRect)
        -> (origin: CGPoint, size: CGSize, needsResize: Bool)
    {
        let plan = slopdesk_window_placement(
            windowSize.width, windowSize.height, // RAW CGSize fields
            displayBounds.origin.x, displayBounds.origin.y, // RAW CGRect origin
            displayBounds.width, displayBounds.height, // CG-standardized (abs)
        )
        return (
            CGPoint(x: plan.origin_x, y: plan.origin_y),
            CGSize(width: plan.width, height: plan.height),
            plan.needs_resize,
        )
    }

    /// True when `size` fits inside `bounds` (within a ½-pt tolerance). Used after the AX move to
    /// confirm the window actually shrank to fit the VD — an app that refuses/clamps the resize
    /// leaves an oversized window, which must NOT be reported as a successful 2× move (the capture
    /// crop would exceed the framebuffer and the client's input mapping would desync).
    /// `bounds.width`/`.height` are CG-standardized (abs); `size` is raw.
    public static func fits(_ size: CGSize, within bounds: CGRect) -> Bool {
        slopdesk_window_fits(size.width, size.height, bounds.width, bounds.height)
    }
}

/// Moves a target window onto a display via Accessibility (feature #1 — put the remoted window on
/// the HiDPI virtual display so it renders at real 2× backing).
///
/// Every line of the sequence — resolve the window, read its frame, plan, shrink, move, read back,
/// roll back on refusal — is `slopdesk-ffi`'s `ax` module now, because the ORDER is load-bearing and
/// one of the steps was already a golden-pinned Rust rule. What is left here is the call and the
/// logging, and the logging is the only reason these are not one-liners.
///
/// Restoration: ``moveWindowOntoDisplay`` reports the window's PRE-move global frame in its result so
/// the caller (the daemon's ``WindowParkingManager``) can put the window back exactly where it was
/// when the pane closes / the daemon shuts down / the VD is torn down — otherwise the user's real
/// window is left shrunk + stranded on the (physically invisible) VD.
public enum WindowPlacement {
    private static let log = Logger(subsystem: "slopdesk.video.host", category: "WindowPlacement")

    /// Outcome of a successful park: the window's ACHIEVED point size on the VD (read back from AX —
    /// the window may have clamped the requested shrink) and its PRE-move global frame (for restore).
    public struct MoveResult: Equatable, Sendable {
        public let achievedSize: CGSize
        public let originalFrame: CGRect
        public init(achievedSize: CGSize, originalFrame: CGRect) {
            self.achievedSize = achievedSize
            self.originalFrame = originalFrame
        }
    }

    /// Move the window `windowID` (owned by `pid`) fully onto `displayID`. `nil` on ANY failure —
    /// window not found, original frame unreadable, AX write failed, hung app, OR the app refused
    /// the shrink so the window still overhangs the VD; on the last two the door already rolled the
    /// window back before answering, so the caller's 1× fallback captures it cleanly in place.
    ///
    /// ⚠️ **GUI-ONLY + TCC:** needs the Accessibility grant (as watcher/injector do).
    @discardableResult
    public static func moveWindowOntoDisplay(
        windowID: CGWindowID,
        pid: pid_t,
        displayID: CGDirectDisplayID,
    ) -> MoveResult? {
        var parked = SlopDeskAxPark()
        guard slopdesk_ax_park_window(windowID, pid, displayID, &parked) else {
            log.error("move window \(windowID): AX park refused for pid \(pid) — staying 1× in place")
            return nil
        }
        let achieved = CGSize(width: parked.achieved_width, height: parked.achieved_height)
        log
            .notice(
                "moved window \(windowID) onto display \(displayID) size \(Int(achieved.width))×\(Int(achieved.height))pt",
            )
        return MoveResult(achievedSize: achieved, originalFrame: HostDisplays.rect(parked.original))
    }

    /// Put the window `windowID` (owned by `pid`) BACK to a previously-saved global frame. Inverse of
    /// ``moveWindowOntoDisplay``. Call before the VD is destroyed so the original display still
    /// exists.
    @discardableResult
    public static func restoreWindow(windowID: CGWindowID, pid: pid_t, toFrame frame: CGRect) -> Bool {
        guard slopdesk_ax_restore_window(windowID, pid, HostDisplays.record(frame)) else {
            log.error("restore window \(windowID): AX window not found (pid \(pid))")
            return false
        }
        log
            .notice(
                "restored window \(windowID) to (\(Int(frame.origin.x)),\(Int(frame.origin.y))) size \(Int(frame.width))×\(Int(frame.height))pt",
            )
        return true
    }

    /// Un-minimize the window `windowID` (owned by `pid`) so the WindowServer paints it again — a
    /// minimized window is never rendered, so SCK capture of one streams nothing (the mint-time
    /// rescue, ``OffScreenWindowMintRescue``, calls this before resolving the window for capture).
    /// Read-then-write: a window that is not minimized is left untouched (`.notMinimized`).
    public static func deminiaturizeWindow(windowID: CGWindowID, pid: pid_t) -> DeminiaturizeOutcome {
        switch slopdesk_ax_deminiaturize(windowID, pid) {
        case Int32(SLOPDESK_AX_DEMINIATURIZE_RESTORING):
            log.notice("deminiaturized window \(windowID) (pid \(pid)) for capture")
            return .restoring
        case Int32(SLOPDESK_AX_DEMINIATURIZE_NOT_MINIMIZED):
            return .notMinimized
        default:
            log.error("deminiaturize window \(windowID): AX window not found or refused (pid \(pid))")
            return .failed
        }
    }
}
#endif
